core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1986,101 +1965,125 @@ private[controller] class ControllerStats extends KafkaMetricsGroup {
 
 sealed trait ControllerEvent {
   def state: ControllerState
+  def preempt(): Unit
 }
 
 case object ControllerChange extends ControllerEvent {
-  override def state = ControllerState.ControllerChange
+  override def state: ControllerState = ControllerState.ControllerChange
+  override def preempt(): Unit = {}
```

Comment:
Can we define a default implementation on the trait rather than overriding all these with empty functions?

Reply:
@mumrah one of the issues was that people are able to create controller events without implementing preemption. In most cases, this is fine because most events don't have callbacks. But for events that do have callbacks, someone can forget to implement preemption and this won't be caught during compile time. This is to force an implementation of preemption which can be empty when designing a new event.

Reply:
@jeffkbkim I think, we can add default implementation on the trait and add a comment saying that to implement preempt() method for events which have callbacks.  I hope, PR reviewers will catch if someone forget to implement preemption methods in future